### PR TITLE
Dev th create return book

### PR DESCRIPTION
**Pull request by trevorhuis-smoothstack**
I accidentally did the return book and check out book getting lost in learning node/express. I have good status codes, but am not sure if my error handling is where it should be.

### CODE REVIEW

**Comment by trevorhuis-smoothstack** [comment]
I have my previous pr in this one.

**Comment by trevorhuis-smoothstack** [comment]
I need to fix my template literals to avoid sql injection.

File: controller/borrowerController.js [hunk 1]
```diff
@@ -1,21 +1,55 @@
...
+  let goodReq = checkLoanCheckOutBody(body, res);
+  if (!goodReq) return;
+
+  service.checkOutBook(body, res);
 });
 
 module.exports = routes;
+
+let checkLoanReturnBody = (body, res) => {
+  if (
```
Comment (sean-darsie) [inline controller/borrowerController.js]: try to put this error handling in the service layer if possible.

File: service/borrowerService.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
+let dao = require("../dao/borrowerDAO.js");
+let moment = require("moment");
```
Comment (justin-obrien-smoothstack) [inline service/borrowerService.js]: const would be more fitting than let for these two, but of course it works either way.

File: service/borrowerService.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      return;
+    }
+
+    // Check if loan exists
+    if (result.length < 1) {
+      res.status(404).send("We could not find that loan in the database. ");
+      return;
+    } else if (result.length > 0 && result[0].dateIn !== null) {
+      // Check if the book is turned in
+      res.status(406).send("That book is already returned. ");
```
Comment (justin-obrien-smoothstack) [inline service/borrowerService.js]: 400 would probably be the best status code to use here. 406 indicates that the server can't meet the requirement given by the Accept header of the request.

File: service/borrowerService.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        loan.dueDate = moment(dueDate).format("YYYY-MM-DD HH:mm:ss");
+        return dao.checkOutBook(loan, result[0].noOfCopies - 1);
+      } else
+        res
+          .status(400)
+          .send("There are no copies of that book in our branch.  ");
+    })
+    // Return the checked out book to the user
+    .then((result) => {
+      res.status(200);
```
Comment (justin-obrien-smoothstack) [inline service/borrowerService.js]: 201 is probably the best status code here.

File: controller/borrowerController.js (hunk 1 above)
Comment (justin-obrien-smoothstack) [reply controller/borrowerController.js]: I'm not 100% sure, but I don't think this error handling is even necessary. The service layer catches cases where the loan isn't in the database, the borrower doesn't exit, or there are no copies (which also covers the cases where the book or branch doesn't exist). I assume if one of these values is undefined, the DAO will substitute the string 'undefined' into the SELECT query which will yield no results and be caught as an error by the service.

**Review by justin-obrien-smoothstack — CHANGES_REQUESTED** [review]
Pretty much everything looks good.

I didn't know it was possible to chain together response.status().send() like that, so that's convenient.

I've found that it's not necessary to backslash-escape quotation marks inside strings that are delimited in backticks.

The only changes I'd definitely request are the couple of status codes I pointed out.

**Review by sean-darsie — APPROVED** [review]
looks good. Could consider cleaning up the dao the way justin did by having a write method that takes care of db queries that use transactions

**Pull request merged**